compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -1068,26 +1068,50 @@ object Types {
      *  Exception (if `-YexplicitNulls` is set): if this type is a nullable union (i.e. of the form `T | Null`),
      *  then the top-level union isn't widened. This is needed so that type inference can infer nullable types.
      */
-    def widenUnion(implicit ctx: Context): Type = widen match {
-      case tp @ OrType(tp1, tp2) =>
-        if (ctx.explicitNulls && tp.isNullableUnion) {
-          // Don't widen `T|Null`, since otherwise we wouldn't be able to infer nullable unions.
-          val OrType(leftTpe, nullTpe) = tp.normNullableUnion
-          OrType(leftTpe.widenUnion, nullTpe)
-        } else {
-          ctx.typeComparer.lub(tp1.widenUnion, tp2.widenUnion, canConstrain = true) match {
-            case union: OrType => union.join
-            case res => res
+    def widenUnion(implicit ctx: Context): Type = {
+      def normalizeRightOrNull(tp: OrType, lhs: Type, rhs: Type): Type = lhs match {
+        case tpp @ OrType(tpp1, tpp2) =>
+          if (tpp2.isJavaNullType) tpp
+          else tpp.derivedOrType(tpp1, rhs)
+        case tpp =>
+          tp.derivedOrType(tpp, rhs)
+      }
+
+      widen match {
+        case tp @ OrType(lhs, rhs) =>
+          def defaultJoin(tp1: Type, tp2: Type) =
+            ctx.typeComparer.lub(tp1, tp2, canConstrain = true) match {
+              case union: OrType => union.join
+              case res => res
+            }
+          if (ctx.explicitNulls) {
+            // Don't widen `T|Null`, since otherwise we wouldn't be able to infer nullable unions.
```

Comment:
As it stands, this is hard to follow because there are so many cases.

I'd propose the following variant:

```Scala
// Test for nullable union that assumes the type has already been normalized.
def isNullableUnionFast(tp: Type): Boolean = tp match {
  // Don't destruct `orTpe` to avoid the method call to `unapply`.
  case orTpe: OrType if orTpe.tp2.isNullType => true
  case _ => false
}

if (rhs.isNullType) ensureNullableUnion(lhs.widenUnion, tp)
else if (lhs.isNullType) ensureNullableUnion(rhs.widenUnion, tp)
else {
  val lhsWiden = lhs.widenUnion
  val rhsWiden = rhs.widenUnion
  val tmpRes = ctx.typeComparer.lub(lhsWiden, rhsWiden, canConstrain = true) match {
    case union: OrType => union.join
    case res => res
  }
  if (isNullableUnionFast(lhsWiden) || isNullableUnionFast(rhsWiden)) ensureNullableUnion(tmpRes)
  else tmpRes
}
```

Reply:
Same problem here, I think we need to track the Null type